src/endpoint-aspnetcore/Endpoint.AspNetCore/EndpointApplicationBuilder.cs

```diff
@@ -0,0 +1,127 @@
+        }
+
+        _ = app.UseRouter(routeBuilder.Build());
+
+        if (app is ISwaggerBuilder swaggerBuilder)
+        {
+            _ = swaggerBuilder.Use(EndpointSwaggerConfigurator.Configure<TEndpoint>);
+        }
+
+        return app;
+    }
+
+    private static async Task InvokeAsync(HttpContext context, IEndpointInvokeSupplier endpoint, string operationId)
+    {
+        var request = new EndpointRequest(
+            operationId: operationId,
+            headers: context.Request.Headers?.SelectMany(GetValues).ToArray(),
+            queryParameters: context.Request.Query?.SelectMany(GetValues).ToArray(),
+            routeValues: context.Request.RouteValues?.Select(MapRouteValue).ToArray(),
+            userClaims: context.User?.Claims?.Select(MapClaimValue).ToArray(),
+            body: context.Request.Body);
+
+        var response = await endpoint.InvokeAsync(request, context.RequestAborted);
+        await context.WriteResponseAsync(response);
+
+        static IEnumerable<KeyValuePair<string, string?>> GetValues(KeyValuePair<string, StringValues> pair)
+        {
+            foreach (var value in pair.Value)
+            {
+                yield return new(pair.Key, value);
+            }
+        }
+
+        static KeyValuePair<string, string?> MapRouteValue(KeyValuePair<string, object?> pair)
+            =>
+            new(
+                pair.Key, pair.Value?.ToString());
+
+        static KeyValuePair<string, string?> MapClaimValue(Claim claim)
+            =>
+            new(claim.Type, claim.Value);
+    }
+
+    private static Task WriteResponseAsync(this HttpContext context, EndpointResponse response)
+    {
+        var httpResponse = context.Response;
+        httpResponse.StatusCode = response.StatusCode;
+
+        foreach (var header in response.Headers)
+        {
+            if (string.IsNullOrEmpty(header.Value))
+            {
+                continue;
+            }
+
+            httpResponse.AddHeader(header.Key, header.Value);
+        }
+
+        if ((response.Body is null) || (response.StatusCode is StatusCodes.Status204NoContent))
+        {
+            return Task.CompletedTask;
+        }
+
+        return httpResponse.WriteBodyAsync(response.Body, context.RequestAborted);
+    }
+
+    private static void AddHeader(this HttpResponse httpResponse, string name, string value)
+    {
+        if (httpResponse.Headers.ContainsKey(name) is false)
+        {
+            httpResponse.Headers.Append(name, value);
+            return;
+        }
+
+        var headerValue = httpResponse.Headers[name];
+        httpResponse.Headers[name] = StringValues.Concat(headerValue, value);
+    }
+
+    private static async Task WriteBodyAsync(this HttpResponse httpResponse, Stream body, CancellationToken cancellationToken)
+    {
+        var buffer = new Memory<byte>(new byte[body.Length]);
+
+        _ = await body.ReadAsync(buffer, cancellationToken);

```

Comment:
The body here is `EndpointResponse.Body` of `Stream` type.
`Stream.Length` may throw `NotSupportedException`.

Suggest to implement a more robust reading in a cycle by small segments (and the last segment may contain less effective bytes read than the segment size).

Or, if for the **Early** we expect a particular `Stream` implementation has working `Stream.Length`, consider following reading, to emphasize that we expect the particular bytes read specified in `Stream.Length`:

```suggestion
        _ = await body.ReadExactlyAsync(buffer, cancellationToken);
```